### PR TITLE
Fix elfutils +debuginfod installation issue

### LINKED ISSUE #36710: Installation issue: elfutils
### Steps to reproduce the issue

```console
$ spack spec -I spack spec -I elfutils@0.189%gcc@11.2.0~bzip2+debuginfod+nls~xz~zstd build_system=autotools arch=linux-centos7-x86_64
Input spec
--------------------------------
 -   elfutils@0.189%gcc@11.2.0~bzip2+debuginfod+nls~xz~zstd build_system=autotools arch=linux-centos7-x86_64

Concretized
--------------------------------
 -   elfutils@0.189%gcc@11.2.0~bzip2+debuginfod+nls~xz~zstd build_system=autotools arch=linux-centos7-x86_64
[+]      ^curl@7.88.1%gcc@11.2.0~gssapi~ldap+libidn2~librtmp~libssh~libssh2~nghttp2 build_system=autotools libs=shared,static tls=openssl arch=linux-centos7-x86_64
[+]          ^libidn2@2.3.4%gcc@11.2.0 build_system=autotools arch=linux-centos7-x86_64
[+]              ^libunistring@1.1%gcc@11.2.0 build_system=autotools arch=linux-centos7-x86_64
[+]          ^openssl@1.1.1t%gcc@11.2.0~docs~shared build_system=generic certs=mozilla arch=linux-centos7-x86_64
[+]              ^ca-certificates-mozilla@2023-01-10%gcc@11.2.0 build_system=generic arch=linux-centos7-x86_64
[+]              ^perl@5.36.0%gcc@11.2.0+cpanm+open+shared+threads build_system=generic arch=linux-centos7-x86_64
[+]                  ^berkeley-db@18.1.40%gcc@11.2.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-centos7-x86_64
[+]                  ^gdbm@1.23%gcc@11.2.0 build_system=autotools arch=linux-centos7-x86_64
[+]          ^pkgconf@1.8.0%gcc@11.2.0 build_system=autotools arch=linux-centos7-x86_64
[+]      ^gettext@0.21.1%gcc@11.2.0+bzip2+curses+git~libunistring+libxml2+tar+xz build_system=autotools arch=linux-centos7-x86_64
[+]          ^bzip2@1.0.8%gcc@11.2.0~debug~pic+shared build_system=generic arch=linux-centos7-x86_64
[+]          ^libiconv@1.17%gcc@11.2.0 build_system=autotools libs=shared,static arch=linux-centos7-x86_64
[+]          ^libxml2@2.10.3%gcc@11.2.0~python build_system=autotools arch=linux-centos7-x86_64
[+]          ^ncurses@6.4%gcc@11.2.0~symlinks+termlib abi=none build_system=autotools arch=linux-centos7-x86_64
[+]          ^tar@1.34%gcc@11.2.0 build_system=autotools zip=pigz arch=linux-centos7-x86_64
[+]              ^pigz@2.7%gcc@11.2.0 build_system=makefile arch=linux-centos7-x86_64
[+]          ^xz@5.4.1%gcc@11.2.0~pic build_system=autotools libs=shared,static arch=linux-centos7-x86_64
[+]      ^libarchive@3.6.2%gcc@11.2.0+iconv build_system=autotools compression=bz2lib,lz4,lzma,lzo2,zlib,zstd crypto=mbedtls libs=shared,static programs=none xar=libxml2 arch=linux-centos7-x86_64
[+]          ^lz4@1.9.4%gcc@11.2.0 build_system=makefile libs=shared,static arch=linux-centos7-x86_64
[+]          ^lzo@2.10%gcc@11.2.0 build_system=autotools libs=shared,static arch=linux-centos7-x86_64
[+]          ^mbedtls@2.28.2%gcc@11.2.0+pic build_system=makefile build_type=Release libs=static arch=linux-centos7-x86_64
[+]          ^zstd@1.5.4%gcc@11.2.0+programs build_system=makefile compression=none libs=shared,static arch=linux-centos7-x86_64
[+]      ^libmicrohttpd@0.9.70%gcc@11.2.0~https build_system=autotools arch=linux-centos7-x86_64
[+]      ^m4@1.4.19%gcc@11.2.0+sigsegv build_system=autotools patches=9dc5fbd,bfdffa7 arch=linux-centos7-x86_64
[+]          ^diffutils@3.8%gcc@11.2.0 build_system=autotools arch=linux-centos7-x86_64
[+]          ^libsigsegv@2.13%gcc@11.2.0 build_system=autotools arch=linux-centos7-x86_64
[+]      ^sqlite@3.40.1%gcc@11.2.0+column_metadata+dynamic_extensions+fts~functions+rtree build_system=autotools arch=linux-centos7-x86_64
[+]          ^readline@8.2%gcc@11.2.0 build_system=autotools patches=bbf97f1 arch=linux-centos7-x86_64
[+]      ^zlib@1.2.13%gcc@11.2.0+optimize+pic+shared build_system=makefile arch=linux-centos7-x86_64
```

### Error message

<details><summary>Error message</summary><pre>

```
==> Installing elfutils-0.189-rflm2fga7fem45hsysiex7jmzs54wjoj
==> No binary for elfutils-0.189-rflm2fga7fem45hsysiex7jmzs54wjoj found: installing from source
==> Using cached archive: /home/gitlab-runner/Build/release/spack/var/spack/cache/_source-cache/archive/39/39bd8f1a338e2b7cd4abc3ff11a0eddc6e690f69578a57478d8179b4148708c8.tar.bz2
==> Ran patch() for elfutils
==> elfutils: Executing phase: 'autoreconf'
==> elfutils: Executing phase: 'configure'
==> Error: ProcessError: Command exited with status 1:
    '/tmp/gitlab-runner/spack-stage/spack-stage-elfutils-0.189-rflm2fga7fem45hsysiex7jmzs54wjoj/spack-src/configure' '--prefix=/cvmfs/sw.hsf.org/spackages7/elfutils/0.189/x86_64-centos7-gcc11.2.0-opt/rflm2' '--without-bzlib' '--without-lzma' '--without-zstd' '--with-zlib=/cvmfs/sw.hsf.org/spackages7/zlib/1.2.13/x86_64-centos7-gcc11.2.0-opt/2wmsk' 'LDFLAGS=-Wl,--no-as-needed -L/cvmfs/sw.hsf.org/spackages7/gettext/0.21.1/x86_64-centos7-gcc11.2.0-opt/4mu7e/lib -lintl' '--enable-debuginfod' '--enable-libdebuginfod'

1 error found in build log:
     133    checking for pkg-config... (cached) /bin/pkg-config
     134    checking pkg-config is at least version 0.9.0... yes
     135    checking for libmicrohttpd... yes
     136    checking for oldlibmicrohttpd... no
     137    checking for sqlite3... yes
     138    checking for libarchive... no
  >> 139    configure: error: dependencies not found, use --disable-debuginfod to disable.

```

</pre></details>
### Information on your system

```
* **Spack:** 0.20.0.dev0
* **Python:** 3.6.8
* **Platform:** linux-centos7-cascadelake
* **Concretizer:** clingo
```

### Additional information

@mwkrentel 

[spack-build-out.txt](https://github.com/spack/spack/files/11182937/spack-build-out.txt)
[spack-build-env.txt](https://github.com/spack/spack/files/11182936/spack-build-env.txt)

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

### PR DESCRIPTION
**Pull request by tmadlener**
Fixes #36710

- Add new versions for `libmicrohttpd`. Also add an older one to make `elfutils` configure with `+debuginfod`
- Require `libarchive~iconv` in order to avoid another configure error.
- Make `elfutils` depend on `pkgconfig` explicitly to avoid picking up the system one.

### CODE REVIEW

File: var/spack/repos/builtin/packages/elfutils/package.py
Comment (tmadlener) [inline var/spack/repos/builtin/packages/elfutils/package.py]: https://github.com/spack/spack/issues/36710#issuecomment-1502117314:
> I can confirm that the problem initially presented itself with libarchive@3.6.2. Using the previous version allows elfutils to build. 

 maybe we should make this version dependent then?

Comment (G-Ragghianti) [reply var/spack/repos/builtin/packages/elfutils/package.py]: Sure.

**Comment by G-Ragghianti** [comment]
This resolves my issues in building gdb.

File: var/spack/repos/builtin/packages/elfutils/package.py
Comment (tmadlener) [reply var/spack/repos/builtin/packages/elfutils/package.py]: Made `libarchive@3.6.2 +iconv` a `conflict`. Building against `libarchive@3.5.2+iconv` works for me as well, and the concretizer is smart enough to figure out `libarchive@3.6.2 ~iconv` by default.

**Comment by tmadlener** [comment]
@hainest would you be OK with this temporary fix until uptream libarchive is fixed?

**Comment by hainest** [comment]
@tmadlener This works for me. Thanks for working on it!

**Review by hainest — DISMISSED** [review]

**Comment by mwkrentel** [comment]
This is, at best a temporary solution until libarchive sorts itself out, right?

There was an active discussion of this in the libarchive pull requests
in Dec/Jan.  The problem centers around iconv being provided by glibc
and how to handle that case (where no extra lib is used) vs a separate
libiconv.

https://github.com/libarchive/libarchive/pull/1825
https://github.com/libarchive/libarchive/pull/1817
https://github.com/libarchive/libarchive/pull/1813
https://github.com/libarchive/libarchive/pull/1812

All of the above are short but a bit different for how they solve the
problem.  People argued forcefully for and against different patches.

The last one, 1825 got merged after 3.6.2 was released, although some
people argued that it was the wrong solution.

Anyone know enough about pkgconfig to say what is the "right" solution
here (I don't).  As I said, 1825 is now in the libarchive code base,
so maybe go with that?

**Comment by tmadlener** [comment]
Yes, I would also consider this a temporary solution. I am also no expert in any way in pkgconfig, so I can't weigh in either on which would be the proper way to fix this.

I have however entirely missed that they merged a PR to "fix" this (for a definition of fix). We could also try to apply that patch to `libarchive@3.6.2` via spack and see if that let's us remove the conflict here.

**Comment by mwkrentel** [comment]
@tmadlener @G-Ragghianti @scheibelp 
I've been experimenting with possible fixes, including 1825 from
libarchive and I'm going around in circles and getting flustrated.
I believe there is a better fix, but probably not one I'll come
up with in a reasonable time.

So, with the understanding that this is a temporary workaround until
we get the "right fix," I suggest we should merge this and move on.

@scheibelp Why is CI failing?  The patch includes:

```
conflicts("libarchive@3.6.2 +iconv", when="+debuginfod")
```

And the error message is:

```
Run . share/spack/setup-env.sh
PKG-DIRECTIVES: 1 issue found
1. elfutils: wrong variant in "conflicts" directive
    the variant 'iconv' does not exist
    in /home/runner/work/spack/spack/var/spack/repos/builtin/packages/elfutils/package.py
Error: Process completed with exit code 1.
```

Is spack assigning `+iconv` to elfutils instead of libarchive?
Do I need to write this as `^libarchive +iconv` instead of `libarchive +iconv`
to force `iconv` to go with libarchive??

Note: libarchive has a `+iconv` variant, but elfutils does not.

**Review by mwkrentel — DISMISSED** [review]

**Comment by G-Ragghianti** [comment]
> Do I need to write this as `^libarchive +iconv` instead of `libarchive +iconv`

Yes I think you just need this change.

**Comment by mwkrentel** [comment]
@tmadlener I checked with Massimiliano at the Wed spack meeting.  Yes, you need
to add the carat `^` to libarchive, that will fix the CI failure.

```
conflicts("^libarchive@3.6.2 +iconv", when="+debuginfod")
```

Without the `^`, this is interpreted as `libarchive@3.6.2` AND `+iconv`,
that is, two separate things.

File: var/spack/repos/builtin/packages/elfutils/package.py
Comment (tmadlener) [inline var/spack/repos/builtin/packages/elfutils/package.py]:
```suggestion
    conflicts("^libarchive@3.6.2 +iconv", when="+debuginfod")
```

**Comment by tmadlener** [comment]
Thanks for the hint. Fixed now, and CI seems happier.

**Review by mwkrentel — DISMISSED** [review]

**Comment by tmadlener** [comment]
@mwkrentel @hainest I think this needs approval again, after fixing the merge conflict

**Review by hainest — APPROVED** [review]

**Review by mwkrentel — APPROVED** [review]

**Comment by tmadlener** [comment]
@scheibelp is there anything still missing for this to get merged?

**Review by scheibelp — APPROVED** [review]

**Comment by scheibelp** [comment]
Thanks!

**Pull request merged**
